### PR TITLE
Some parser and AST cleanups

### PR DESCRIPTION
**Pull request by nnethercote**
Things I found while looking closely at this code.

r? @petrochenkov

### CODE REVIEW

File: compiler/rustc_builtin_macros/src/assert.rs [hunk 1]
```diff
@@ -58,7 +58,7 @@ pub fn expand_assert<'cx>(
                 path: panic_path(),
                 args: P(DelimArgs {
                     dspan: DelimSpan::from_single(call_site_span),
-                    delim: MacDelimiter::Parenthesis,
+                    delim: token::Delimiter::Parenthesis,
```
Comment (petrochenkov) [inline compiler/rustc_builtin_macros/src/assert.rs:61]:
```suggestion
                    delim: Delimiter::Parenthesis,
```
Could you avoid 3-segment paths for enum variants and import things instead (unless there are some name conflicts)?

**Comment by petrochenkov** [comment]
r=me after addressing https://github.com/rust-lang/rust/pull/114353#discussion_r1281347662 and fixing doc comments.
@rustbot author

**Comment by nnethercote** [comment]
Thank you for the fast review. I addressed the review comments and fixed the doc comments.

@bors r=petrochenkov rollup

**Comment by nnethercote** [comment]
I fixed the rustfmt formatting error.

@bors r=petrochenkov

**Pull request merged**